Docs: Improve the Spark Structured Streaming jobs (#8203)
apache/iceberg: docs/spark-structured-streaming.md

To write values from streaming query to Iceberg table, use `DataStreamWriter`:

```scala
data.writeStream
    .format("iceberg")
    .outputMode("append")
    .trigger(Trigger.ProcessingTime(1, TimeUnit.MINUTES))
    .option("checkpointLocation", checkpointPath)

>     .toTable("database.table_name")

Reviewer: Which versions of Flink does this work for? Is it new and that's why we didn't show it before? May want a note about compatibility if that's the case.

Author: Ah, that's a good call. Since 3.1: https://github.com/apache/spark/blob/e4114f67e12a235b4784fcbfa6f6ba9b44a5e715/sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala#L267

enable fanout writer to eliminate the requirement.

```scala
data.writeStream
    .format("iceberg")
    .outputMode("append")
    .trigger(Trigger.ProcessingTime(1, TimeUnit.MINUTES))
    .option("fanout-enabled", "true")
    .option("checkpointLocation", checkpointPath)
    .toTable("database.table_name")
```

Fanout writer opens the files per partition value and doesn't close these files till write task is finished. This option is discouraged for batch writing, as explicit sort against output rows is cheap for batch workloads.

## Maintenance for streaming tables

Streaming writes can create new table versions quickly, creating lots of table metadata to track those versions.
Maintaining metadata by tuning the rate of commits, expiring old snapshots, and automatically cleaning up metadata files
is highly recommended.

### Tune the rate of commits

> Having high rate of commits produces data files, manifests, and snapshots which leads to additional maintenance. We encourage having trigger interval 1 minute at minimum, and increase the interval if needed.

Reviewer: Nit: I like to avoid personal pronouns like "we" or "I" in documentation because it isn't clear who the speaker is (unlike a blog post, when it is clear that "I" is the author).

Author: I fully agree there, thanks!
```suggestion
Having a high rate of commits produces data files, manifests, and snapshots which leads to additional maintenance. It is recommended to have a trigger interval of 1 minute at the minimum and increase the interval if needed.
```

enable fanout writer to eliminate the requirement.

```scala
data.writeStream
    .format("iceberg")
    .outputMode("append")
    .trigger(Trigger.ProcessingTime(1, TimeUnit.MINUTES))
    .option("fanout-enabled", "true")
    .option("checkpointLocation", checkpointPath)
    .toTable("database.table_name")
```

> Fanout writer opens the files per partition value and doesn't close these files till write task is finished. This option is discouraged for batch writing, as explicit sort against output rows is cheap for batch workloads.

Reviewer 2: ```suggestion
Fanout writer opens the files per partition value and doesn't close these files till the write task finishes. Avoid using the fanout writer for batch writing, as explicit sort against output rows is cheap for batch workloads.
```

https://learn.microsoft.com/en-us/style-guide/grammar/verbs#active-and-passive-voice

## Maintenance for streaming tables

Streaming writes can create new table versions quickly, creating lots of table metadata to track those versions.
Maintaining metadata by tuning the rate of commits, expiring old snapshots, and automatically cleaning up metadata files
is highly recommended.

### Tune the rate of commits

Having a high rate of commits produces data files, manifests, and snapshots which leads to additional maintenance. It is recommended to have a trigger interval of 1 minute at the minimum and increase the interval if needed.

The triggers section in [Structured Streaming Programming Guide](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#triggers)
documents how to configure the interval.

### Expire old snapshots

> Each batch written to a table produces a new snapshot, that is tracked in table metadata until they are expired. Snapshots accumulate quickly with frequent commits, so it is highly recommended that tables written by streaming queries are [regularly maintained](../maintenance#expire-snapshots). [Snapshot expiration](../spark-procedures/#expire_snapshots) is the procedure of removing the metadata and any data files that are no longer needed. By default, the procedure will expire the snapshots older than 5 days. 

Reviewer 2: ```suggestion
Each batch written to a table produces a new snapshot. Iceberg tracks snapshots in table metadata until they are expired. Snapshots accumulate quickly with frequent commits, so it is highly recommended that tables written by streaming queries are [regularly maintained](../maintenance#expire-snapshots). [Snapshot expiration](../spark-procedures/#expire_snapshots) is the procedure of removing the metadata and any data files that are no longer needed. By default, the procedure will expire the snapshots older than five days. 
```
https://learn.microsoft.com/en-us/style-guide/numbers#numerals-vs-words

### Compacting data files

The amount of data written from a streaming process is typically small, which can cause the table metadata to track lots of small files. [Compacting small files into larger files](../maintenance#compact-data-files) reduces the metadata needed by the table, and increases query efficiency. Iceberg and Spark [comes with the `rewrite_data_files` procedure](../spark-procedures/#rewrite_data_files).

### Rewrite manifests

> To optimize write latency on streaming workload, Iceberg may write the new snapshot with a "fast" append that does not automatically compact manifests.

Reviewer 2: ```suggestion
To optimize write latency on a streaming workload, Iceberg can write the new snapshot with a "fast" append that does not automatically compact manifests.
```

> This could lead lots of small manifest files. Manifests can be [rewritten to optimize queries and to compact](../maintenance#rewrite-manifests). Iceberg and Spark [comes with the `rewrite_manifests` procedure](../spark-procedures/#rewrite_manifests).

Reviewer 2: ```suggestion
This could lead lots of small manifest files. Iceberg can [rewrite the number of manifest files to improve query performance](../maintenance#rewrite-manifests). Iceberg and Spark [come with the `rewrite_manifests` procedure](../spark-procedures/#rewrite_manifests).
```
